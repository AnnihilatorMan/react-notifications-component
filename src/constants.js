export const NOTIFICATION_BASE_CLASS = "notification-item";

export const CONTAINER = {
  BOTTOM_LEFT: "bottom-left",
  BOTTOM_RIGHT: "bottom-right",
  TOP_LEFT: "top-left",
  TOP_RIGHT: "top-right"
};

export const INSERTION = {
  TOP: "top",
  BOTTOM: "bottom"
};

export const NOTIFICATION_TYPE = {
  SUCCESS: "success",
  DANGER: "danger",
  INFO: "info",
  DEFAULT: "default",
  WARNING: "warning"
};

export const NOTIFICATION_STAGE = {
  // used for both sliding and animation at the same time
  SLIDING_ANIMATION_EXIT: "SLIDING_ANIMATION_EXIT",

  // used for onTouchMove/onTouchEnd to slide and animate
  TOUCH_SLIDING_ANIMATION_EXIT: "TOUCH_SLIDING_ANIMATION_EXIT",

  // used by API call to remove notification
  REMOVAL: "REMOVAL"
};
