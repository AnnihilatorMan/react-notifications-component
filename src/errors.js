export default {
  // dismiss icon option
  DISMISS_ICON_CLASS: "className property of dismissIcon option is required",
  DISMISS_ICON_CONTENT: "content property of dismissIcon option is required",
  DISMISS_ICON_STRING: "className property of dismissIcon option must be a String",
  DISMISS_ICON_INVALID: "content property of dismissIcon option must be a valid React element",

  // animations
  ANIMATION_IN: "animationIn option must be an array",
  ANIMATION_OUT: "animationOut option must be an array",

  // dismiss
  DISMISS_REQUIRED: "duration property of dismiss option is required",
  DISMISS_NUMBER: "duration property of dismiss option must be a Number",
  DISMISS_POSITIVE: "duration property of dismiss option must be a positive Number",

  // title
  TITLE_STRING: "title option must be a String.",

  // message
  MESSAGE_REQUIRED: "message option is required",
  MESSAGE_STRING: "message option must be a String",

  // type
  TYPE_REQUIRED: "type option is required",
  TYPE_STRING: "type option must be a String",
  TYPE_NOT_EXISTENT: "type option not found",

  // container
  CONTAINER_REQUIRED: "container option is required",
  CONTAINER_STRING: "container option must be a String",

  // dismissable
  DISMISSABLE_CLICK_BOOL: "click property of dismissable option must be a Boolean",
  DISMISSABLE_TOUCH_BOOL: "touch property of dismissable option must be a Boolean",

  // width
  WIDTH_NUMBER: "width option must be a Number",

  // insert
  INSERT_STRING: "insert option must be a String",

  // transition
  TRANSITION_DURATION_NUMBER: "duration property of transition option must be a Number",
  TRANSITION_CUBICBEZIER_NUMBER: "cubicBezier property of transition option must be a String",
  TRANSITION_DELAY_NUMBER: "delay property of transition option must be a Number",

  // custom types
  TYPE_NOT_FOUND: "custom type not found"
};
