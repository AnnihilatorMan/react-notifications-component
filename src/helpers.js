import React from "react";
import {
  handleTouchSlidingAnimationExit,
  handleSlidingAnimationExit
} from "./stage-helpers";

import {
  NOTIFICATION_BASE_CLASS,
  CONTAINER,
  INSERTION,
  NOTIFICATION_TYPE,
  NOTIFICATION_STAGE
} from "./constants";

import {
  isNullOrUndefined,
  isString,
  isNumber,
  isBoolean,
  isArray
} from "./utils";

import ERROR from "./errors";

export function isBottomContainer(container) {
  return (
    container === CONTAINER.BOTTOM_LEFT ||
    container === CONTAINER.BOTTOM_RIGHT
  );
}

export function isTopContainer(container) {
  return (
    container === CONTAINER.TOP_LEFT ||
    container === CONTAINER.TOP_RIGHT
  );
}

export function shouldNotificationHaveSliding(notification) {
  return (
    // slide DOWN if container is top and insertion is at top
    (notification.insert === INSERTION.TOP && isTopContainer(notification.container)) ||
    // slide UP if container is bottom and insertion is at bottom
    (notification.insert === INSERTION.BOTTOM && isBottomContainer(notification.container))
  );
}

export function cssWidth(width) {
  return width ? `${width}px` : undefined;
}

function htmlClassesForExistingType(type) {
  switch (type.toLowerCase()) {
    case NOTIFICATION_TYPE.DEFAULT:
      return [NOTIFICATION_BASE_CLASS, "notification-default"];
    case NOTIFICATION_TYPE.SUCCESS:
      return [NOTIFICATION_BASE_CLASS, "notification-success"];
    case NOTIFICATION_TYPE.DANGER:
      return [NOTIFICATION_BASE_CLASS, "notification-danger"];
    case NOTIFICATION_TYPE.WARNING:
      return [NOTIFICATION_BASE_CLASS, "notification-warning"];
    case NOTIFICATION_TYPE.INFO:
      return [NOTIFICATION_BASE_CLASS, "notification-info"];
    default:
      throw new Error(`Type '${type}' is not valid nor user-defined`);
  }
}

export function getHtmlClassesForType(notification) {
  const {
    // option set by user
    type,
    // array value containing all custom types
    userDefinedTypes
  } = notification;

  if (notification.content) {
    // return only base class if type is not defined
    return [NOTIFICATION_BASE_CLASS];
  } else if (isNullOrUndefined(userDefinedTypes)) {
    // existing type
    return htmlClassesForExistingType(type);
  }

  // look for custom type if any defined
  const foundType = userDefinedTypes.find(q => q.name === type);

  if (!foundType) {
    // custom type not found, throw error
    throw new Error(ERROR.TYPE_NOT_FOUND);
  }

  // append base class to html classes
  return [NOTIFICATION_BASE_CLASS].concat(foundType.htmlClasses);
}

export function getNotificationsForMobileView(notifications) {
  const {
    TOP_LEFT,
    TOP_RIGHT,
    BOTTOM_LEFT,
    BOTTOM_RIGHT
  } = CONTAINER;

  const bottom = [];
  const top = [];

  notifications.forEach((notification) => {
    const container = notification.container.toLowerCase();
    switch (container) {
      case TOP_LEFT:
      case TOP_RIGHT:
        top.push(notification);
        break;
      case BOTTOM_LEFT:
      case BOTTOM_RIGHT:
        bottom.push(notification);
        break;
      default:
        // throw error just in case validation didn't work properly
        throw new Error(`Container ${notification.container} is not valid`);
    }
  });

  return { top, bottom };
}

export function getNotificationsForEachContainer(notifications) {
  const topLeft = [];
  const topRight = [];
  const bottomLeft = [];
  const bottomRight = [];

  notifications.forEach((notification) => {
    const {
      TOP_LEFT,
      TOP_RIGHT,
      BOTTOM_LEFT,
      BOTTOM_RIGHT
    } = CONTAINER;

    const container = notification.container.toLowerCase();
    switch (container) {
      case TOP_LEFT:
        topLeft.push(notification);
        break;
      case TOP_RIGHT:
        topRight.push(notification);
        break;
      case BOTTOM_LEFT:
        bottomLeft.push(notification);
        break;
      case BOTTOM_RIGHT:
        bottomRight.push(notification);
        break;
      default:
        // throw error just in case validation didn't work properly
        throw new Error(`Container ${notification.container} is not valid`);
    }
  });

  return {
    topLeft,
    topRight,
    bottomLeft,
    bottomRight
  };
}

export function getInitialSlidingState({ notification, isFirstNotification }) {
  // no sliding needed for first notification in container
  const hasSliding = shouldNotificationHaveSliding(notification) && !isFirstNotification;
  const state = {};

  // set default classes for animated element
  state.animatedElementClasses = getHtmlClassesForType(notification);
  state.rootElementStyle = {
    height: "0",
    marginBottom: 0,
    overflow: "hidden",
    width: cssWidth(notification.width)
  };

  if (hasSliding) {
    // hide content by toggling visibility while sliding
    state.animatedElementClasses.push("invisible");
  } else if (notification.animationIn && notification.animationIn.length > 0) {
    // add user defined notification classes if sliding will not occur
    notification.animationIn.forEach(item => state.animatedElementClasses.push(item));
  }

  state.hasSliding = hasSliding;
  return state;
}

export function handleStageTransition(notification, state) {
  let animatedElementClasses;
  let rootElementStyle;

  const {
    animatedElementClasses: stateAnimatedElementClasses,
    rootElementStyle: stateRootStyle,
    currentX,
    startX
  } = state;

  if (notification.stage === NOTIFICATION_STAGE.TOUCH_SLIDING_ANIMATION_EXIT) {
    return handleTouchSlidingAnimationExit(notification, currentX, startX);
  } else if (notification.stage === NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT) {
    return handleSlidingAnimationExit(notification);
  } else if (notification.resized) {
    // window got resized, do not apply animations
    rootElementStyle = stateRootStyle;
    animatedElementClasses = getHtmlClassesForType(notification);
  } else {
    // use values from state
    rootElementStyle = stateRootStyle;
    animatedElementClasses = stateAnimatedElementClasses;
  }

  return {
    rootElementStyle,
    animatedElementClasses
  };
}

export function getCubicBezierTransition(
  duration = 500,
  cubicBezier = "linear",
  delay = 0,
  property = "height"
) {
  return `${duration}ms ${property} ${cubicBezier} ${delay}ms`;
}

export function hasFullySwiped(diffX) {
  const swipeLength = Math.abs(diffX);
  const requiredSwipeLength = (40 / 100) * window.innerWidth;

  return swipeLength >= requiredSwipeLength;
}

export function getRootHeightStyle(notification, scrollHeight) {
  return {
    height: `${scrollHeight}px`,
    width: cssWidth(notification.width),
    transition: getCubicBezierTransition(
      notification.slidingExit.duration,
      notification.slidingExit.cubicBezier,
      notification.slidingExit.delay
    )
  };
}

export function getIconHtmlContent(notification, onClickHandler) {
  // use icon defined by user
  if (notification.dismissIcon) {
    return (
      <div
        className={notification.dismissIcon.className}
        onClick={onClickHandler}
      >
        {notification.icon}
      </div>
    );
  }

  // use standard icon
  return (
    <div
      className="notification-close"
      onClick={onClickHandler}
    >
      <span>&times;</span>
    </div>
  );
}

export function validateDismissIconOption(dismissIcon) {
  if (isNullOrUndefined(dismissIcon)) return;

  const { className, content } = dismissIcon;

  if (!className) {
    // className is required
    throw new Error(ERROR.DISMISS_ICON_CLASS);
  } else if (!isString(className)) {
    // must be an array value
    throw new Error(ERROR.DISMISS_ICON_STRING);
  } else if (!content) {
    // content is required
    throw new Error(ERROR.DISMISS_ICON_CONTENT);
  } else if (!React.isValidElement(content)) {
    // must be a valid react element
    throw new Error(ERROR.DISMISS_ICON_INVALID);
  }
}

export function validateAnimationIn(animationIn) {
  if (isNullOrUndefined(animationIn)) {
    return [];
  } if (!isArray(animationIn)) {
    // must be an array
    throw new Error(ERROR.ANIMATION_IN);
  }

  return animationIn;
}

export function validateAnimationOut(animationOut) {
  if (isNullOrUndefined(animationOut)) {
    return [];
  } else if (!isArray(animationOut)) {
    // must be an array
    throw new Error(ERROR.ANIMATION_OUT);
  }

  return animationOut;
}

export function validateTimeoutDismissOption(dismiss) {
  // skip if option is not defined
  if (isNullOrUndefined(dismiss)) return;

  if (isNullOrUndefined(dismiss.duration)) {
    throw new Error(ERROR.DISMISS_REQUIRED);
  } else if (!isNumber(dismiss.duration)) {
    // must be a number
    throw new Error(ERROR.DISMISS_NUMBER);
  } else if (dismiss.duration < 0) {
    // duration must be positive
    throw new Error(ERROR.DISMISS_POSITIVE);
  }
}

export function validateTransition(transition, defaults) {
  const { duration, cubicBezier, delay } = defaults;
  const transitionOptions = transition || {};

  if (isNullOrUndefined(transitionOptions.duration)) {
    // set default duration
    transitionOptions.duration = duration;
  }

  if (isNullOrUndefined(transitionOptions.cubicBezier)) {
    // set default timing function
    transitionOptions.cubicBezier = cubicBezier;
  }

  if (isNullOrUndefined(transitionOptions.delay)) {
    // set default delay
    transitionOptions.delay = delay;
  }

  if (!isNumber(transitionOptions.duration)) {
    // throw if duration is NaN
    throw new Error(ERROR.TRANSITION_DURATION_NUMBER);
  } else if (!isString(transitionOptions.cubicBezier)) {
    // throw if cubicBezier is not a String
    throw new Error(ERROR.TRANSITION_CUBICBEZIER_NUMBER);
  } else if (!isNumber(transitionOptions.delay)) {
    // throw if duration is NaN
    throw new Error(ERROR.TRANSITION_DELAY_NUMBER);
  }

  return transitionOptions;
}

export function validateTitle(notification) {
  const { content, title } = notification;

  // option not required if content is defined
  if (content) return;

  if (title === null || title === undefined) {
    // title is required
    throw new Error(ERROR.TITLE_REQUIRED);
  } else if (!isString(title)) {
    // title must be a String
    throw new Error(ERROR.TITLE_STRING);
  }
}

export function validateMessage(notification) {
  const { content, message } = notification;

  // option not required if content is defined
  if (content) return;

  if (message === null || message === undefined) {
    throw new Error(ERROR.MESSAGE_REQUIRED);
  } else if (!isString(message)) {
    throw new Error(ERROR.MESSAGE_STRING);
  }
}

export function validateType(notification, userDefinedTypes) {
  const { content, type } = notification;

  if (content) {
    // option not required if content is defined
    return undefined;
  } else if (isNullOrUndefined(type)) {
    // type is required
    throw new Error(ERROR.TYPE_REQUIRED);
  } else if (!isString(type)) {
    // type must be a String
    throw new Error(ERROR.TYPE_STRING);
  } else if (
    isNullOrUndefined(userDefinedTypes) &&
    type !== NOTIFICATION_TYPE.SUCCESS &&
    type !== NOTIFICATION_TYPE.DANGER &&
    type !== NOTIFICATION_TYPE.INFO &&
    type !== NOTIFICATION_TYPE.DEFAULT &&
    type !== NOTIFICATION_TYPE.WARNING
  ) {
    throw new Error(ERROR.TYPE_NOT_EXISTENT);
  }

  return type.toLowerCase();
}

export function validateContainer(container) {
  if (isNullOrUndefined(container)) {
    // container is required
    throw new Error(ERROR.CONTAINER_REQUIRED);
  } else if (!isString(container)) {
    // container must be a String
    throw new Error(ERROR.CONTAINER_STRING);
  }

  return container.toLowerCase();
}

export function validateDismissable(dismissable) {
  if (isNullOrUndefined(dismissable)) {
    return {
      // set dissmisable by click as default
      click: true,
      // set dismissable on swipe as default
      touch: true
    };
  }

  const option = dismissable;

  if (isNullOrUndefined(option.click)) {
    // set dissmisable by click as default
    option.click = true;
  }

  if (isNullOrUndefined(option.touch)) {
    // set dismissable on swipe as default
    option.touch = true;
  }

  if (!isBoolean(option.click)) {
    // must be boolean
    throw new Error(ERROR.DISMISSABLE_CLICK_BOOL);
  } else if (!isBoolean(option.touch)) {
    // must be boolean
    throw new Error(ERROR.DISMISSABLE_TOUCH_BOOL);
  }

  return option;
}

export function validateInsert(insert) {
  if (isNullOrUndefined(insert)) {
    // default to top if not defined
    return "top";
  } else if (!isString(insert)) {
    // must be a string value (top|bottom)
    throw new Error(ERROR.INSERT_STRING);
  }

  return insert;
}

export function validateWidth(width) {
  if (!isNumber(width)) {
    // width must be a valid Number
    throw new Error(ERROR.WIDTH_NUMBER);
  }

  return width;
}

export function validateUserDefinedTypes(notification, definedTypes) {
  const { content, type } = notification;

  // no need to validate type if content is defined
  if (content) return undefined;

  if (
    type === NOTIFICATION_TYPE.SUCCESS ||
    type === NOTIFICATION_TYPE.DANGER ||
    type === NOTIFICATION_TYPE.INFO ||
    type === NOTIFICATION_TYPE.DEFAULT ||
    type === NOTIFICATION_TYPE.WARNING ||
    isNullOrUndefined(definedTypes)
  ) return undefined;

  // search for custom type in array
  if (!definedTypes.find(p => p.name === type)) {
    // custom type not found, throw exception
    throw new Error(ERROR.TYPE_NOT_FOUND);
  }

  return definedTypes;
}

function getRandomId() {
  return Math.random();
}

export function getNotificationOptions(options, userDefinedTypes) {
  const notification = options;
  const {
    insert,
    container,
    dismissIcon,
    animationIn,
    animationOut,
    slidingEnter,
    slidingExit,
    touchSlidingBack,
    touchSlidingExit,
    dismissable,
    dismiss,
    width
  } = notification;

  // for now we'll use Math.random for id
  notification.id = getRandomId();

  // validate notification's title
  validateTitle(notification);

  // validate notification's body message
  validateMessage(notification);

  // validate notification's type
  notification.type = validateType(notification, userDefinedTypes);

  if (userDefinedTypes && !notification.content) {
    // validate user defined types if any
    notification.userDefinedTypes = validateUserDefinedTypes(notification, userDefinedTypes);
  }

  // validate notification's container
  notification.container = validateContainer(container);

  // set default insertion
  notification.insert = validateInsert(insert);

  // set default dismissable options
  notification.dismissable = validateDismissable(dismissable);

  // assert `dismissIcon` matches rules
  validateDismissIconOption(dismissIcon);

  // validate `animationIn` array
  notification.animationIn = validateAnimationIn(animationIn);

  // validate `animationOut` array
  notification.animationOut = validateAnimationOut(animationOut);

  if (!isNullOrUndefined(width)) {
    // set width option only if defined
    notification.width = validateWidth(width);
  }

  // defaults
  const slidingEnterDefaults = { duration: 600, cubicBezier: "linear", delay: 0 };
  const slidingExitDefaults = { duration: 600, cubicBezier: "linear", delay: 0 };
  const swipeBackDefaults = { duration: 600, cubicBezier: "ease-in", delay: 0 };
  const swipeExitDefaults = { duration: 600, cubicBezier: "ease-in", delay: 0 };
  const swipeCompleteDefaults = { duration: 300, cubicBezier: "ease-in", delay: 0 };

  // set defaults for sliding enter transition
  notification.slidingEnter = validateTransition(slidingEnter, slidingEnterDefaults);

  // set defaults for sliding exit transition
  notification.slidingExit = validateTransition(slidingExit, slidingExitDefaults);

  // set defaults for sliding back on touchEnd/touchCancel
  notification.touchSlidingBack = validateTransition(touchSlidingBack, swipeBackDefaults);

  // let it empty object if undefined
  notification.touchSlidingExit = notification.touchSlidingExit || {};

  // set defaults for swipe transition when swipe is complete
  notification.touchSlidingExit.swipe = validateTransition(
    touchSlidingExit.swipe || {},
    swipeExitDefaults
  );

  // set defaults for fade transition when swipe is complete
  notification.touchSlidingExit.fade = validateTransition(
    touchSlidingExit.fade || {},
    swipeCompleteDefaults
  );

  // assert `dismiss` matches rules
  validateTimeoutDismissOption(dismiss);

  return notification;
}
