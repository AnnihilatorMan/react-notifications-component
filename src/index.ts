import { Container } from 'src/components/Container'
import Store from 'src/store'
export default Container

import {
  iContainerProps,
  iContainerState,
  iNotification,
  NotificationTitleMessage,
  NotificationContent,
  iTransition,
  iTouchTransition,
  iNotificationDismiss,
  iNotificationCustomType,
  iNotificationProps,
  iNotificationState,
  iNotificationParentStyle,
  iStore,
  iNotificationStoreParams,
  NOTIFICATION_CONTAINER,
  NOTIFICATION_TYPE,
  NOTIFICATION_INSERTION,
  NOTIFICATION_REMOVAL_SOURCE
} from './types'

export { Store as store } // otherwise it's breaking change
export { Container as ReactNotifications }

export {
  iContainerProps,
  iContainerState,
  iNotification,
  NotificationTitleMessage,
  NotificationContent,
  iTransition,
  iTouchTransition,
  iNotificationDismiss,
  iNotificationCustomType,
  iNotificationProps,
  iNotificationState,
  iNotificationParentStyle,
  iStore,
  iNotificationStoreParams,
  NOTIFICATION_CONTAINER,
  NOTIFICATION_TYPE,
  NOTIFICATION_INSERTION,
  NOTIFICATION_REMOVAL_SOURCE
}
