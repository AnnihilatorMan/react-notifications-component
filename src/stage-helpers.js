import {
  getHtmlClassesForType,
  getCubicBezierTransition,
  cssWidth
} from './helpers';

export function touchSwipeTransition(notification) {
  const { swipe } = notification.touchSlidingExit;

  return getCubicBezierTransition(
    swipe.duration,
    swipe.cubicBezier,
    swipe.delay,
    "left"
  );
}

export function touchFadeTransition(notification) {
  const { fade } = notification.touchSlidingExit;

  return getCubicBezierTransition(
    fade.duration,
    fade.cubicBezier,
    fade.delay,
    "opacity"
  );
}

export function slidingExitTransition(notification) {
  return getCubicBezierTransition(
    notification.slidingExit.duration,
    notification.slidingExit.cubicBezier,
    notification.slidingExit.delay
  );
}

export function getChildStyleForTouchTransitionExit(notification, currentX, startX) {
  const horizontalLimit = window.innerWidth * 2;
  const touchSwipe = exports.touchSwipeTransition(notification);
  const touchFade = exports.touchFadeTransition(notification);

  // set `opacity` and `left` to pull out notification from screen
  return {
    opacity: 0,
    position: "relative",
    left: `${currentX - startX >= 0 ? horizontalLimit : -horizontalLimit}px`,
    transition: `${touchSwipe}, ${touchFade}`
  };
}

export function handleTouchSlidingAnimationExit(notification, currentX, startX) {
  // set current html classes
  const animatedElementClasses = getHtmlClassesForType(notification);
  // set opacity and left to pull-out notification
  const childElementStyle = getChildStyleForTouchTransitionExit(notification, currentX, startX);
  // sliding out transition
  const slidingTransition = exports.slidingExitTransition(notification);

  return {
    childElementStyle,
    animatedElementClasses,
    // slide to height 0
    rootElementStyle: {
      height: 0,
      marginBottom: 0,
      transition: slidingTransition,
      width: cssWidth(notification.width)
    },
  };
}

export function handleSlidingAnimationExit(notification) {
  const { animationOut } = notification;
  const animatedElementClasses = getHtmlClassesForType(notification);

  if (animationOut) {
    // add animation classes defined by user if any
    animationOut.forEach(item => animatedElementClasses.push(item));
  }

  return {
    // slide element to height 0
    rootElementStyle: {
      height: 0,
      marginBottom: 0,
      transition: slidingExitTransition(notification),
      width: cssWidth(notification.width)
    },
    animatedElementClasses
  };
}
