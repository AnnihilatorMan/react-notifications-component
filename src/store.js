import { parseNotification } from './utils/helpers';
import { validateTransition, validators } from './utils/validators';

function Store() {
  this.types = null;
  this.add = () => {};

  this.addNotification = (notification) => {
    const { types } = this;

    if (process.env.NODE_ENV === 'development') {
      const transitions = ['slidingEnter','slidingExit', 'touchRevert', 'touchSlidingExit'];
      transitions.forEach((transition) => validateTransition(notification, transition));
      validators.forEach((validator) => validator(notification, types));
    }

    return this.add(parseNotification(notification, types));
  };
  
  this.removeNotification = () => {};
  this.register = ({ addNotification, removeNotification, types }) => {
    this.add = addNotification;
    this.removeNotification = removeNotification;
    this.types = types;
  };

  return this;
}

export default new Store();