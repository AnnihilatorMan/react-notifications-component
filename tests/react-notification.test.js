import React from "react";
import ReactNotification from "src/react-notification";
import Enzyme, { mount } from "enzyme";
import React16Adapter from "enzyme-adapter-react-16";
import sinon from "sinon";
import notificationMock from "tests/mocks/notification.mock";
import { NOTIFICATION_STAGE } from "src/constants";

Enzyme.configure({
  // react 16 adapter
  adapter: new React16Adapter()
});

describe("Notification component", () => {
  let clock;
  let component;

  // arrow function helpers
  const getNotificationMock = (edits = {}) => Object.assign({}, notificationMock, edits);
  const instance = () => component.instance();
  const state = () => component.state();

  beforeEach(() => {
    // set fake timer
    clock = sinon.useFakeTimers();
  });

  afterEach(() => {
    // restore fake timer
    clock.restore();

    if (component) {
      // unmount
      component.unmount();
    }
  });

  it("onTransitionEnd updates state with needed CSS classes", () => {
    component = mount(
      <ReactNotification notification={getNotificationMock({ animationIn: ["html5", "css3"] })} />
    );

    // manually call function
    component.instance().onTransitionEnd();
    // expect correct state update
    expect(component.state).toMatchSnapshot();

    component = mount(
      <ReactNotification notification={getNotificationMock({ animationIn: null })} />
    );
    // manually call function
    component.instance().onTransitionEnd();
    // expect correct state update
    expect(component.state).toMatchSnapshot();
  });

  it("onTouchSmartSlidingEnd returns if target node is NOT the root node", () => {
    component = mount(
      <ReactNotification notification={getNotificationMock({
        stage: NOTIFICATION_STAGE.TOUCH_SLIDING_ANIMATION_EXIT
      })} />
    );

    component.find(".notification-item-child").simulate("transitionend");
    // expect endOfSmartSliding to be false
    expect(instance().endOfSmartSliding).toBe(false);
  });

  it("onTouchSmartSlidingEnd toggles removal", () => {
    const toggleRemoval = jest.fn();
    component = mount(
      <ReactNotification toggleRemoval={toggleRemoval} notification={getNotificationMock({
        stage: NOTIFICATION_STAGE.TOUCH_SLIDING_ANIMATION_EXIT
      })} />
    );

    component.find(".notification-item-root").simulate("transitionend")
    // expect endOfSmartSliding to be true
    expect(instance().endOfSmartSliding).toBe(true);

    component.find(".notification-item-root").simulate("transitionend")
    // expect toggleRemoval to be called
    expect(toggleRemoval.mock.calls.length).toBe(1);
  });

  it("onSmartSlidingEnd toggles removal for undefined/empty animationOut", () => {
    const toggleRemoval = jest.fn();
    component = mount(
      <ReactNotification toggleRemoval={toggleRemoval} notification={getNotificationMock({
        stage: NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT,
        animationOut: null
      })} />
    );

    // trigger transitionend event
    component.find(".notification-item-root").simulate("transitionend");
    expect(toggleRemoval.mock.calls.length).toBe(1);

    component = mount(
      <ReactNotification toggleRemoval={toggleRemoval} notification={getNotificationMock({
        stage: NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT,
        animationOut: []
      })} />
    );

    // trigger transitionend event
    component.find(".notification-item-root").simulate("transitionend");
    expect(toggleRemoval.mock.calls.length).toBe(2);
  });

  it("onSmartSlidingEnd toggles removal on 2nd function call", () => {
    const toggleRemoval = jest.fn();
    component = mount(
      <ReactNotification toggleRemoval={toggleRemoval} notification={getNotificationMock({
        stage: NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT
      })} />
    );

    // trigger transitionend event
    component.find(".notification-item-root").simulate("transitionend");
    expect(toggleRemoval.mock.calls.length).toBe(0);

    // trigger transitionend event
    component.find(".notification-item-root").simulate("transitionend");
    expect(toggleRemoval.mock.calls.length).toBe(1);
  });

  it("componentDidMount sets initial state", () => {
    const spy = jest.spyOn(ReactNotification.prototype, "componentDidMount");

    component = mount(
      <ReactNotification notification={getNotificationMock({ resized: false })} />
    );

    expect(spy.mock.calls.length).toBe(1);
    expect(state()).toMatchSnapshot();

    component = mount(
      <ReactNotification notification={getNotificationMock({ resized: true })} />
    );

    expect(spy.mock.calls.length).toBe(2);
    expect(state()).toMatchSnapshot();
  });

  it("timeout handler skips setting state if stage is REMOVAL/EXIT", () => {
    const notification = getNotificationMock({
      dismiss: { duration: 200 },
      stage: NOTIFICATION_STAGE.TOUCH_SLIDING_ANIMATION_EXIT
    });

    component = mount(<ReactNotification notification={notification} />);

    clock.tick(400);

    expect(component.state()).toMatchSnapshot();
  });

  it("timeout handler updates state", () => {
    const notification = getNotificationMock({
      dismiss: { duration: 100 },
      stage: undefined
    });
    const toggleTimeoutRemoval = jest.fn();

    component = mount(<ReactNotification
      notification={notification}
      toggleTimeoutRemoval={toggleTimeoutRemoval}
    />);

    clock.tick(400);

    expect(toggleTimeoutRemoval.mock.calls.length).toBe(1);
    expect(component.state()).toMatchSnapshot();
  });

  it("component sets timeout for duration > 0", () => {
    // add dismiss option with duration set
    const notification = getNotificationMock({ dismiss: { duration: 2000 } });

    // mount
    component = mount(<ReactNotification notification={notification} />);

    // expect timeoutId to have been set
    expect(instance().timeoutId).toBeDefined();
  });

  it("component does not set timeout for duration = 0", () => {
    // add dismiss option with duration set
    const notification = getNotificationMock({ dismiss: { duration: 0 } });

    // mount
    component = mount(<ReactNotification notification={notification} />);

    expect(instance().timeoutId).toBeUndefined();
  });

  it("component does not set timeout for duration < 0", () => {
    // add dismiss option with duration set
    const notification = getNotificationMock({ dismiss: { duration: -1 } });

    // mount
    component = mount(<ReactNotification notification={notification} />);

    expect(instance().timeoutId).toBeUndefined();
  });

  it("component does not set timeout for undefined dismiss option", () => {
    const notification = getNotificationMock({ dismiss: {} });

    // mount
    component = mount(<ReactNotification notification={notification} />);

    expect(instance().timeoutId).toBeUndefined();
  });

  it("clicks notification and calls onNotificationClick prop", () => {
    const onClickHandler = jest.fn();
    const spy = jest.spyOn(ReactNotification.prototype, "onNotificationClick");

    component = mount(<ReactNotification
      notification={notificationMock}
      onClickHandler={onClickHandler}
    />);

    // trigger click
    component.find(".notification-item").simulate("click");

    // expect instance onClickHandler to be called
    expect(spy.mock.calls.length).toBe(1);

    // tick 100ms for requestAnimation
    clock.tick(100);

    // expect onClickHandler prop to have been called
    expect(onClickHandler.mock.calls.length).toBe(1);
  });
});