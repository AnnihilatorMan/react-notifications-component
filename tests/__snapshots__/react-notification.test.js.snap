// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component timeout handler skips setting state if stage is REMOVAL/EXIT 1`] = `
Object {
  "animatedElementClasses": Array [
    "notification-item",
    "notification-success",
    "animated",
    "fadeIn",
  ],
  "hasSliding": true,
  "rootElementStyle": Object {
    "height": "0px",
    "transition": "600ms height cubic-bezier(0.215, 0.61, 0.355, 1) 0ms",
    "width": undefined,
  },
}
`;

exports[`Notification component timeout handler updates state 1`] = `
Object {
  "animatedElementClasses": Array [
    "notification-item",
    "notification-success",
    "animated",
    "fadeIn",
  ],
  "hasSliding": true,
  "rootElementStyle": Object {
    "height": "0px",
    "transition": "600ms height cubic-bezier(0.215, 0.61, 0.355, 1) 0ms",
    "width": undefined,
  },
}
`;
